Added events to Summoning and Counterspell

clcment446 opened the pull request:
Added the "PlayerSummonsCreature" event, used in summoning spells. 
Added the "CounterSpellEvent" event, used whenever someone uses counterspell.

### Contributor Liscence Agreement
[Full CLA](https://github.com/iron431/Irons-Spells-n-Spellbooks/blob/1.19.2/CLA.md)
*Contributors: Please "X" in the following box to acknowledge our CLA*
- [X] I have read and agree to the CLA for Iron's Spells and Spellbooks which allows me to retain my ownership in the code submitted while granting the repository owner legal rights to use my contribution.

clcment446 committed:
ISS CreatureSummonedEvent

Added the "PlayerSummonsCreature" event, used in summoning spells.
Added the "CounterSpellEvent" event, used whenever someone uses counterspell.

iron431 commented on src/main/java/io/redspace/ironsspellbooks/spells/blood/RaiseDeadSpell.java:
```diff
@@ -98,6 +100,8 @@ public void onCast(Level world, int spellLevel, LivingEntity entity, CastSource
             undead.setYRot(entity.getYRot());
             undead.setOldPosAndRot();
+            var event = NeoForge.EVENT_BUS.post(new PlayerSummonsCreature(entity, undead, this.spellId, spellLevel));
+            undead = (Monster) event.getCreature();
```
this, and all the other uses of the event, use an unchecked cast. the assignment is unnecessary and could be be inlined to `level.addFreshEntity(event.getCreature());`

iron431 commented on src/main/java/io/redspace/ironsspellbooks/api/events/PlayerSummonsCreature.java:
```diff
@@ -0,0 +1,39 @@
+import net.minecraft.world.entity.LivingEntity;
+import net.neoforged.neoforge.event.entity.living.LivingEvent;
+
+public class PlayerSummonsCreature extends LivingEvent {
```
this is not a player event. a better name could be something like "SpellSummonEvent"

clcment446 committed:
Fix unsafe Casting in caused by summoning event

- PlayerSummonEvents -> SpellSummonEvent

- SpellSummonEvents now works with a generic to avoid potential class casting issues.
